packages/web3-core-requestmanager/src/index.js

```diff
@@ -123,11 +123,13 @@ RequestManager.prototype.setProvider = function (provider, net) {
             });
         });
 
-        // notify all subscriptions about the close condition
+        // notify all subscriptions about bad close conditions and unsubscribe
         this.provider.on('close', function close(event) {
-            _this.subscriptions.forEach(function (subscription) {
-                subscription.callback(new Error('CONNECTION ERROR: The connection got closed with the close code `' + event.code + '` and the following reason string `' + event.reason + '`'));
-            });
+            if (![1000, 1001].includes(event.code) || event.wasClean === false) {
```

Comment:
- Do those rules also work with an IPC provider? 
- Shouldn't such additional error rules get handled in the related provider? 
- Aren't we hiding errors with those changes?

Reply:
> Do those rules also work with an IPC provider?
Good question!

> Aren't we hiding errors with those changes?
At the moment we are throwing errors when there aren't any. (See tests). Can you think of a test which would show an error being swallowed here? Agree that's super important...and will think about it more.

> Shouldn't such additional error rules get handled in the provider classes? 
I think that would be ideal - but this is where the listener callbacks are processed and this block of code is where the error is already being managed (see existing error). 

What would you propose as an alternative?

Reply:
> Good question!

I've checked the ``net`` module [code on GitHub](https://github.com/nodejs/node/blob/master/lib/net.js) and the documentation of nodejs. The close listener doesn't have all those information passed as we know from the ``WebSocket`` object. The only information you get is a boolean called ``hadError``.

> At the moment we are throwing errors when there aren't any. (See tests). Can you think of a test which would show an error being swallowed here? Agree that's super important...and will think about it more.

I see your point. From a networking perspective, they aren't errors but from a "running-subscription perspective" they are.
We have removed once the error listener because after an error does a ``WebSocket`` ``CloseEvent`` follow. The correct implementation would probably be to have your addition from the ``RequestManager`` moved to the ``_onClose`` listener and to emit the ``WebsocketProvider`` error event if those conditions are fulfilled.  

> What would you propose as an alternative? 

I would move this logic to the ``_onClose`` listener of the ``WebsocketProvider`` with adding of the ``error`` event listener in the ``RequestManager``. The ``_onClose`` listener should then detect if the ``WebSocket`` connection got closed because of an ``Error`` (un-expected/un-clean close) and emit the correct ``WebsocketProvider`` events.

Reply:
I've done some tests with the idea I had here and it sadly doesn't work as I thought it would. The solution you proposed is the only one which is working as expected. One addition we could do is to clean up the ``subscriptions`` map on ``close`` and ``error`` to not have a memory leak.

Reply:
... and it is required to extend the condition to handle the subscriptions with the ``IpcProvider`` nicely as well.

``` javascript 
if ((typeof event === 'boolean' && event) || (![1000, 1001].includes(event.code) || event.wasClean === false)) {
```

Reply:
> they aren't errors but from a "running-subscription perspective" they are.

This is from my point of view sadly something we can't handle in a good way in any case. For example, if the connected node does close the connection in a clean way but it's unexpected behavior for the client. The developer would have to listen to the close event of his provider to handle this. 
One possible solution we could implement is to add a flag to the ``CloseEvent`` to later detect if it was a clean close from the client-side or remote side.  🤔

Reply:
Ok interesting. 

191ae06eb0d595ee07a96f6d58ff3f76d24b6087 adds handling for the IPC case, plus made the test assertions a little more detailed.

35663ec65fe78c18a3914b9956810fc5d5ab9b1d cleans up the event handlers in the tests - am seeing them fire (usually trying to close the server) in a few places after the test is completed.

Reply:
I've just added some comments to improve the code style but anything else looks good! Thanks! :)